history.go

```diff
@@ -0,0 +1,75 @@
+package merkletree
+
+import (
+	"errors"
+
+	"github.com/coniks-sys/libmerkleprefixtree-go/crypto"
+)
+
+var (
+	ErrHistoryExisted = errors.New("[merkletree] trying to override existing history")
+	ErrBadEpoch       = errors.New("[merkletree] next epoch's STR has bad epoch ")
+)
+
+type History struct {
+	tree          *MerkleTree
+	currentSTR    *SignedTreeRoot
+	epochInterval int64
+}
+
+func NewHistory(m *MerkleTree, key crypto.SigningKey, startEp, epInterval int64) *History {
+	h := new(History)
+	h.tree = m
+	h.epochInterval = epInterval
+	h.currentSTR = newSTR(m, startEp, 0, make([]byte, crypto.HashSizeByte), key)
+	return h
+}
+
+func (h *History) UpdateHistory(m *MerkleTree, nextEp int64) error {
```

Comment:
> we have 2 different STRs of each epoch, right?

I don't think so. Let me try to explain and see if you agree.

At `t == 0`, we don't have any STRs, we've just handed out a few temporary bindings. We can set them in the tree, but since we haven't signed the root yet, any lookups just come from the temporary bindings.

At `t == 1`, we take a snapshot (clone) of the tree as it is and sign the root. We shift that STR in an array. We continue to hand out temporary bindings and set them in the tree. Now, when a lookup comes, we take `array[0]` and lookup in there. If we don't find it, we search our temporary bindings as a fallback.

At `t == 2`, we repeat the same as `1`.

Since the STRs always contain a clone of the tree, we're free to update the tree in between epochs. An epoch as a snapshot of the state at a current time. That's what I mean by it being discreet.

Is that helpful?